File: tests/wasm/test_cfo.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('CFO partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.cfo_js(close, 14, 100.0);
    assert.strictEqual(result.length, close.length);
});

test('CFO accuracy', async () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.cfo;

    const result = wasm.cfo_js(
        close,
        expected.defaultParams.period,
        expected.defaultParams.scalar
    );

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-6,
        "CFO last 5 values mismatch"
    );


    await compareWithRust('cfo', result, 'close', expected.defaultParams);
});

test('CFO default candles', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.cfo_js(close, 14, 100.0);
    assert.strictEqual(result.length, close.length);
});

test('CFO zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.cfo_js(inputData, 0, 100.0);
    }, /Invalid period/);
});

test('CFO period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.cfo_js(dataSmall, 10, 100.0);
    }, /Invalid period/);
});

test('CFO very small dataset', () => {

    const singlePoint = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.cfo_js(singlePoint, 14, 100.0);
    }, /Invalid period|Not enough valid data/);
});

test('CFO empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.cfo_js(empty, 14, 100.0);
    }, /No data provided/);
});

test('CFO reinput', () => {

    const close = new Float64Array(testData.close);


    const firstResult = wasm.cfo_js(close, 14, 100.0);
    assert.strictEqual(firstResult.length, close.length);


    const secondResult = wasm.cfo_js(firstResult, 14, 100.0);
    assert.strictEqual(secondResult.length, firstResult.length);


    if (secondResult.length > 240) {
        for (let i = 240; i < secondResult.length; i++) {
            assert(!isNaN(secondResult[i]), `Found unexpected NaN at index ${i}`);
        }
    }
});

test('CFO NaN handling', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.cfo_js(close, 14, 100.0);
    assert.strictEqual(result.length, close.length);


    let firstValid = 0;
    for (let i = 0; i < close.length; i++) {
        if (!isNaN(close[i])) {
            firstValid = i;
            break;
        }
    }
    const warmupPeriod = firstValid + 14 - 1;


    assertAllNaN(result.slice(0, warmupPeriod), `Expected NaN in warmup period [0:${warmupPeriod})`);


    if (result.length > warmupPeriod) {
        let hasValidValue = false;
        for (let i = warmupPeriod; i < result.length; i++) {
            if (!isNaN(result[i])) {
                hasValidValue = true;
                break;
            }
        }
        assert(hasValidValue, "Expected some valid values after warmup");
    }
});

test('CFO all NaN input', () => {

    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);

    assert.throws(() => {
        wasm.cfo_js(allNaN, 14, 100.0);
    }, /All values are NaN/);
});

test('CFO batch single parameter set', () => {

    const close = new Float64Array(testData.close);


    const batchResult = wasm.cfo_batch_js(
        close,
        14, 14, 0,
        100.0, 100.0, 0
    );


    const singleResult = wasm.cfo_js(close, 14, 100.0);

    assert.strictEqual(batchResult.length, singleResult.length);

    assertArrayClose(batchResult, singleResult, 1e-7, "Batch vs single mismatch");
});

test('CFO batch multiple periods', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const batchResult = wasm.cfo_batch_js(
        close,
        10, 14, 2,
        100.0, 100.0, 0
    );


    assert.strictEqual(batchResult.length, 3 * 100);


    const periods = [10, 12, 14];
    for (let i = 0; i < periods.length; i++) {
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.slice(rowStart, rowEnd);

        const singleResult = wasm.cfo_js(close, periods[i], 100.0);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Period ${periods[i]} mismatch`
        );
    }
});

test('CFO batch metadata', () => {

    const metadata = wasm.cfo_batch_metadata_js(
        10, 14, 2,
        50.0, 150.0, 50.0
    );



    assert.strictEqual(metadata.length, 9 * 2);


    assert.strictEqual(metadata[0], 10);
    assert.strictEqual(metadata[1], 50.0);


    assert.strictEqual(metadata[16], 14);
    assert.strictEqual(metadata[17], 150.0);
});

test('CFO batch full parameter sweep', () => {

    const close = new Float64Array(testData.close.slice(0, 50));

    const batchResult = wasm.cfo_batch_js(
        close,
        10, 12, 2,
        50.0, 100.0, 50.0
    );

    const metadata = wasm.cfo_batch_metadata_js(
        10, 12, 2,
        50.0, 100.0, 50.0
    );


    assert.strictEqual(batchResult.length, 4 * 50);
    assert.strictEqual(metadata.length, 4 * 2);
});

test('CFO batch unified API', () => {

    const close = new Float64Array(testData.close.slice(0, 100));

    const config = {
        period_range: [10, 14, 2],
        scalar_range: [50.0, 150.0, 50.0]
    };

    const result = wasm.cfo_batch(close, config);


    assert(result.values, "Missing values field");
    assert(result.combos, "Missing combos field");
    assert(result.rows, "Missing rows field");
    assert(result.cols, "Missing cols field");


    assert.strictEqual(result.rows, 9);
    assert.strictEqual(result.cols, 100);
    assert.strictEqual(result.values.length, 9 * 100);
    assert.strictEqual(result.combos.length, 9);


    result.combos.forEach(combo => {
        assert(combo.hasOwnProperty('period'), "Combo missing period");
        assert(combo.hasOwnProperty('scalar'), "Combo missing scalar");
    });
});

test('CFO constant values', () => {

    const constant = new Float64Array(50);
    constant.fill(42.0);

    const result = wasm.cfo_js(constant, 14, 100.0);
    assert.strictEqual(result.length, constant.length);



    let foundNonNaN = false;
    for (let i = 0; i < result.length; i++) {
        if (!isNaN(result[i])) {
            foundNonNaN = true;
            assert(Math.abs(result[i]) < 1e-10, `CFO should be ~0 for constant series at index ${i}, got ${result[i]}`);
        }
    }
    assert(foundNonNaN, "Should have at least one non-NaN value");
});

test('CFO linear trend', () => {

    const x = new Float64Array(100);
    for (let i = 0; i < 100; i++) {
        x[i] = i;
    }

    const result = wasm.cfo_js(x, 14, 100.0);
    assert.strictEqual(result.length, x.length);



    let foundNonNaN = false;
    for (let i = 0; i < result.length; i++) {
        if (!isNaN(result[i])) {
            foundNonNaN = true;
            assert(Math.abs(result[i]) < 1e-10, `CFO should be ~0 for linear trend at index ${i}, got ${result[i]}`);
        }
    }
    assert(foundNonNaN, "Should have at least one non-NaN value");
});

test('CFO scalar edge cases', () => {

    const data = new Float64Array([10.0, 20.0, 30.0, 40.0, 50.0]);


    const result1 = wasm.cfo_js(data, 2, NaN);
    let allNaN = true;
    for (let i = 0; i < result1.length; i++) {
        if (!isNaN(result1[i])) {
            allNaN = false;
            break;
        }
    }
    assert(allNaN, "NaN scalar should produce all NaN output");


    const result2 = wasm.cfo_js(data, 2, Infinity);

    for (let i = 1; i < result2.length; i++) {
        assert(isNaN(result2[i]) || !isFinite(result2[i]),
               `Infinite scalar should produce inf/nan at index ${i}`);
    }


    const result3 = wasm.cfo_js(data, 2, 0.0);
    for (let i = 1; i < result3.length; i++) {
        assert.strictEqual(result3[i], 0.0,
                          `Zero scalar should produce zero at index ${i}`);
    }
});

test('CFO negative scalar', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const resultPos = wasm.cfo_js(close, 14, 100.0);
    const resultNeg = wasm.cfo_js(close, 14, -100.0);

    assert.strictEqual(resultPos.length, resultNeg.length);


    for (let i = 0; i < resultPos.length; i++) {
        if (isNaN(resultPos[i]) && isNaN(resultNeg[i])) {
            continue;
        }
        assertClose(-resultPos[i], resultNeg[i], 1e-10,
                   `Negative scalar mismatch at index ${i}`);
    }
});

test('CFO warmup period', () => {

    const close = new Float64Array(testData.close);


    const periods = [5, 10, 14, 20, 30];
    for (const period of periods) {
        const result = wasm.cfo_js(close, period, 100.0);


        let firstValid = 0;
        for (let i = 0; i < close.length; i++) {
            if (!isNaN(close[i])) {
                firstValid = i;
                break;
            }
        }
        const expectedWarmup = firstValid + period - 1;


        for (let i = 0; i < expectedWarmup && i < result.length; i++) {
            assert(isNaN(result[i]),
                  `Period ${period}: Expected NaN at index ${i} (before warmup ${expectedWarmup})`);
        }


        if (expectedWarmup < result.length) {
            assert(!isNaN(result[expectedWarmup]),
                  `Period ${period}: Expected valid value at index ${expectedWarmup}`);
        }
    }
});

test('CFO edge values', () => {



    const dataWithZero = new Float64Array([10.0, 20.0, 30.0, 0.0, 40.0, 50.0, 60.0]);
    const result1 = wasm.cfo_js(dataWithZero, 3, 100.0);
    assert(isNaN(result1[3]), "Expected NaN when current value is 0");


    const dataWithInf = new Float64Array([10.0, 20.0, 30.0, Infinity, 40.0, 50.0, 60.0]);
    const result2 = wasm.cfo_js(dataWithInf, 3, 100.0);
    assert(isNaN(result2[3]), "Expected NaN when current value is inf");


    const dataSmall = new Float64Array([1e-10, 2e-10, 3e-10, 4e-10, 5e-10]);
    const result3 = wasm.cfo_js(dataSmall, 2, 100.0);
    assert.strictEqual(result3.length, dataSmall.length);


    let hasValid = false;
    for (let i = 0; i < result3.length; i++) {
        if (!isNaN(result3[i])) {
            hasValid = true;
            break;
        }
    }
    assert(hasValid, "Should handle very small values");
});










test('CFO batch error handling', () => {

    const close = new Float64Array(testData.close.slice(0, 50));


    assert.throws(() => {
        wasm.cfo_batch(close, { invalid: "config" });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.cfo_batch(close, { period_range: [10, 12, 2] });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.cfo_batch(close, {
            period_range: [0, 0, 0],
            scalar_range: [100.0, 100.0, 0.0]
        });
    }, /Invalid period/);
});

File: tests/python/test_cfo.py
"""
Python binding tests for CFO indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestCfo:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_cfo_partial_params(self, test_data):
        """Test CFO with partial parameters (None values) - mirrors check_cfo_partial_params"""
        close = test_data['close']
        
        
        result = ta_indicators.cfo(close, 14, 100.0)  
        assert len(result) == len(close)
    
    def test_cfo_accuracy(self, test_data):
        """Test CFO matches expected values from Rust tests - mirrors check_cfo_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['cfo']
        
        result = ta_indicators.cfo(
            close,
            period=expected['default_params']['period'],
            scalar=expected['default_params']['scalar']
        )
        
        assert len(result) == len(close)
        
        
        assert_close(
            result[-5:], 
            expected['last_5_values'],
            rtol=1e-6,
            msg="CFO last 5 values mismatch"
        )
        
        
        compare_with_rust('cfo', result, 'close', expected['default_params'])
    
    def test_cfo_default_candles(self, test_data):
        """Test CFO with default parameters - mirrors check_cfo_default_candles"""
        close = test_data['close']
        
        
        result = ta_indicators.cfo(close, 14, 100.0)
        assert len(result) == len(close)
    
    def test_cfo_zero_period(self):
        """Test CFO fails with zero period - mirrors check_cfo_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.cfo(input_data, period=0, scalar=100.0)
    
    def test_cfo_period_exceeds_length(self):
        """Test CFO fails when period exceeds data length - mirrors check_cfo_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.cfo(data_small, period=10, scalar=100.0)
    
    def test_cfo_very_small_dataset(self):
        """Test CFO fails with insufficient data - mirrors check_cfo_very_small_dataset"""
        single_point = np.array([42.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.cfo(single_point, period=14, scalar=100.0)
    
    def test_cfo_empty_input(self):
        """Test CFO fails with empty input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="No data provided"):
            ta_indicators.cfo(empty, period=14, scalar=100.0)
    
    def test_cfo_reinput(self, test_data):
        """Test CFO applied twice (re-input) - mirrors check_cfo_reinput"""
        close = test_data['close']
        
        
        first_result = ta_indicators.cfo(close, period=14, scalar=100.0)
        assert len(first_result) == len(close)
        
        
        second_result = ta_indicators.cfo(first_result, period=14, scalar=100.0)
        assert len(second_result) == len(first_result)
        
        
        if len(second_result) > 240:
            assert not np.any(np.isnan(second_result[240:])), "Found unexpected NaN after warmup period"
    
    def test_cfo_nan_handling(self, test_data):
        """Test CFO handles NaN values correctly - mirrors check_cfo_nan_handling"""
        close = test_data['close']
        
        result = ta_indicators.cfo(close, period=14, scalar=100.0)
        assert len(result) == len(close)
        
        
        first_valid = next((i for i, v in enumerate(close) if not np.isnan(v)), 0)
        warmup_period = first_valid + 14 - 1
        
        
        assert np.all(np.isnan(result[:warmup_period])), f"Expected NaN in warmup period [0:{warmup_period})"
        
        
        if len(result) > warmup_period:
            
            assert not np.all(np.isnan(result[warmup_period:])), "Expected some valid values after warmup"
    
    def test_cfo_streaming(self, test_data):
        """Test CFO streaming matches batch calculation - mirrors check_cfo_streaming"""
        close = test_data['close']
        period = 14
        scalar = 100.0
        
        
        batch_result = ta_indicators.cfo(close, period=period, scalar=scalar)
        
        
        stream = ta_indicators.CfoStream(period=period, scalar=scalar)
        stream_values = []
        
        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)
        
        stream_values = np.array(stream_values)
        
        
        assert len(batch_result) == len(stream_values)
        
        
        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9, 
                        msg=f"CFO streaming mismatch at index {i}")
    
    def test_cfo_batch(self, test_data):
        """Test CFO batch processing - mirrors check_batch_default_row"""
        close = test_data['close']
        
        result = ta_indicators.cfo_batch(
            close,
            period_range=(14, 14, 0),  
            scalar_range=(100.0, 100.0, 0.0)  
        )
        
        assert isinstance(result, dict)
        assert 'values' in result
        assert 'periods' in result
        assert 'scalars' in result
        
        
        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)
        
        
        default_row = result['values'][0]
        expected = EXPECTED_OUTPUTS['cfo']['last_5_values']
        
        
        assert_close(
            default_row[-5:],
            expected,
            rtol=1e-6,
            msg="CFO batch default row mismatch"
        )
    
    def test_cfo_all_nan_input(self):
        """Test CFO with all NaN values"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.cfo(all_nan, period=14, scalar=100.0)

    def test_cfo_batch_sweep(self, test_data):
        """Test CFO batch with parameter sweep"""
        close = test_data['close'][:100]  
        
        result = ta_indicators.cfo_batch(
            close,
            period_range=(10, 20, 5),  
            scalar_range=(50.0, 150.0, 50.0)  
        )
        
        
        assert result['values'].shape[0] == 9
        assert result['values'].shape[1] == len(close)
        assert len(result['periods']) == 9
        assert len(result['scalars']) == 9
        
        
        expected_periods = [10, 10, 10, 15, 15, 15, 20, 20, 20]
        expected_scalars = [50.0, 100.0, 150.0] * 3
        
        np.testing.assert_array_equal(result['periods'], expected_periods)
        np.testing.assert_array_almost_equal(result['scalars'], expected_scalars, decimal=10)

    def test_cfo_with_kernel(self, test_data):
        """Test CFO with different kernel selections"""
        close = test_data['close'][:100]  
        
        
        result_scalar = ta_indicators.cfo(close, 14, 100.0, kernel="scalar")
        assert len(result_scalar) == len(close)
        
        
        result_auto = ta_indicators.cfo(close, 14, 100.0)
        assert len(result_auto) == len(close)
        
        
        assert_close(result_scalar, result_auto, rtol=1e-10, atol=1e-10,
                    msg="Kernel results should match")

    def test_cfo_constant_values(self, test_data):
        """Test CFO with constant input values"""
        constant = np.full(50, 42.0)
        
        result = ta_indicators.cfo(constant, period=14, scalar=100.0)
        assert len(result) == len(constant)
        
        
        
        non_nan_values = result[~np.isnan(result)]
        assert np.all(np.abs(non_nan_values) < 1e-10), "CFO should be ~0 for constant series"

    def test_cfo_linear_trend(self):
        """Test CFO with perfect linear trend"""
        
        x = np.arange(100, dtype=float)
        
        result = ta_indicators.cfo(x, period=14, scalar=100.0)
        assert len(result) == len(x)
        
        
        
        non_nan_values = result[~np.isnan(result)]
        assert np.all(np.abs(non_nan_values) < 1e-10), "CFO should be ~0 for perfect linear trend"

    def test_cfo_scalar_edge_cases(self):
        """Test CFO with edge case scalar values"""
        data = np.array([10.0, 20.0, 30.0, 40.0, 50.0])
        
        
        result = ta_indicators.cfo(data, period=2, scalar=float('nan'))
        assert np.all(np.isnan(result)), "NaN scalar should produce all NaN output"
        
        
        result = ta_indicators.cfo(data, period=2, scalar=float('inf'))
        
        non_warmup = result[1:]  
        assert np.all(np.isnan(non_warmup) | np.isinf(non_warmup)), \
            "Infinite scalar should produce inf/nan output"
        
        
        result = ta_indicators.cfo(data, period=2, scalar=0.0)
        non_warmup = result[1:]
        assert np.all(non_warmup == 0.0), "Zero scalar should produce zero output"
    
    def test_cfo_negative_scalar(self, test_data):
        """Test CFO with negative scalar values (should work)"""
        close = test_data['close'][:100]
        
        
        result_pos = ta_indicators.cfo(close, period=14, scalar=100.0)
        result_neg = ta_indicators.cfo(close, period=14, scalar=-100.0)
        
        assert len(result_pos) == len(result_neg)
        
        
        for i, (pos, neg) in enumerate(zip(result_pos, result_neg)):
            if np.isnan(pos) and np.isnan(neg):
                continue
            assert_close(-pos, neg, rtol=1e-10, atol=1e-10,
                        msg=f"Negative scalar mismatch at index {i}")
    
    def test_cfo_warmup_period(self, test_data):
        """Test CFO warmup period calculation is correct"""
        close = test_data['close']
        
        
        for period in [5, 10, 14, 20, 30]:
            result = ta_indicators.cfo(close, period=period, scalar=100.0)
            
            
            first_valid = next((i for i, v in enumerate(close) if not np.isnan(v)), 0)
            expected_warmup = first_valid + period - 1
            
            
            assert np.all(np.isnan(result[:expected_warmup])), \
                f"Period {period}: Expected NaN before index {expected_warmup}"
            
            
            if expected_warmup < len(result):
                assert not np.isnan(result[expected_warmup]), \
                    f"Period {period}: Expected valid value at index {expected_warmup}"
    
    def test_cfo_edge_values(self):
        """Test CFO with edge case values in data"""
        
        data_with_zero = np.array([10.0, 20.0, 30.0, 0.0, 40.0, 50.0, 60.0])
        result = ta_indicators.cfo(data_with_zero, period=3, scalar=100.0)
        
        
        assert np.isnan(result[3]), "Expected NaN when current value is 0"
        
        
        data_with_inf = np.array([10.0, 20.0, 30.0, np.inf, 40.0, 50.0, 60.0])
        result = ta_indicators.cfo(data_with_inf, period=3, scalar=100.0)
        
        
        assert np.isnan(result[3]), "Expected NaN when current value is inf"
        
        
        data_small = np.array([1e-10, 2e-10, 3e-10, 4e-10, 5e-10])
        result = ta_indicators.cfo(data_small, period=2, scalar=100.0)
        assert len(result) == len(data_small)
        assert not np.all(np.isnan(result)), "Should handle very small values"
    
    def test_cfo_batch_invalid_params(self, test_data):
        """Test CFO batch with invalid parameter ranges"""
        close = test_data['close'][:100]
        
        
        with pytest.raises(ValueError, match="Invalid|period"):
            ta_indicators.cfo_batch(
                close,
                period_range=(20, 10, 1),  
                scalar_range=(100.0, 100.0, 0.0)
            )
        
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.cfo_batch(
                close,
                period_range=(0, 0, 0),  
                scalar_range=(100.0, 100.0, 0.0)
            )
    
    def test_cfo_batch_edge_cases(self, test_data):
        """Test CFO batch with edge case parameter combinations"""
        close = test_data['close'][:50]
        
        
        result = ta_indicators.cfo_batch(
            close,
            period_range=(14, 14, 0),
            scalar_range=(100.0, 100.0, 0.0)
        )
        assert result['values'].shape[0] == 1, "Should have 1 combination"
        
        
        result = ta_indicators.cfo_batch(
            close,
            period_range=(14, 14, 0),
            scalar_range=(-100.0, 100.0, 100.0)  
        )
        assert result['values'].shape[0] == 3, "Should have 3 scalar values"
        
        
        row_neg = result['values'][0]  
        row_pos = result['values'][2]  
        
        for i in range(len(close)):
            if np.isnan(row_neg[i]) and np.isnan(row_pos[i]):
                continue
            if not np.isnan(row_neg[i]) and not np.isnan(row_pos[i]):
                assert_close(-row_neg[i], row_pos[i], rtol=1e-10, atol=1e-10,
                            msg=f"Batch negative scalar mismatch at index {i}")


if __name__ == '__main__':
    pytest.main([__file__, '-v'])